src/wasm/wasm-validator.cpp

```diff
@@ -2084,11 +2099,9 @@ void FunctionValidator::visitRethrow(Rethrow* curr) {
                 Type(Type::unreachable),
                 curr,
                 "rethrow's type must be unreachable");
-  shouldBeSubTypeOrFirstIsUnreachable(
-    curr->exnref->type,
-    Type::exnref,
-    curr->exnref,
-    "rethrow's argument must be exnref type or its subtype");
+  // TODO Validate depth field if it becomes necessary in future. This requires
+  // a separate stack for 'catch', and the current LLVM toolchain only generates
+  // depth 0 for C++ support.
```

Comment:
Would it be worth validating that `depth` is 0 for now? Or do you plan to resolve this TODO in the near future?

Reply:
Validating this requires maintaining a separate stack within the validator, which is not difficult but I'm not very sure if it's worth it at the moment, so I don't have a plan to add that in near future, but if you think it's important we can change priorities. But at the same time I think rejecting values other than 0 can be risky, because binaryen is not guaranteed to take wasm code produced only by our LLVM toolchain.

Reply:
My thinking here is that it would be better to reject a valid program that uses non-zero depths than to accidentally and silently mis-optimize the program because we don't reason about non-zero depths right now. Or do we handle non-zero depths correctly in the optimization passes already?

Reply:
I don't think optimization passes will change the depth field, because they don't have any way of reasoning about that field. But yeah, there can be optimizations that remove or alter try-catch structure that are targeted by `rethrow`. I can't think of any optimization pass that does that right now though.

For one, `Vacuum` [removes](https://github.com/WebAssembly/binaryen/blob/f18c18e01d03d6d293fe3d701408855bbcea58bd/src/passes/Vacuum.cpp#L338-L346) try-catch altogether if the body of `try` does not throw. But this case is OK for `rethrow` because it removes the whole catch body. But there can be other passes that might interfere with `rethrow`, as you pointed out.

I'll reject non-zero depths for now for safe measure.